### ISSUE TITLE
7.9.1 prevent the use of firestore in React Native

### ISSUE THREAD

**Issue by papay0** [firebase/firebase-js-sdk#2667]
Hey,

Just quick FYI to tell you that it looks like 7.9.1 is preventing the use of firestore on React Native.
You can find the detail [here](https://stackoverflow.com/questions/60361519/cant-find-a-variable-atob/60369040#60369040).

Thanks!

**Comment by ImPDP** [comment]
@papay0 did you find any solution this problem? the realtime database is working fine but Firestore has some serious issue with the expo. need help

**Comment by Edweis** [comment]
I can confirm, I faced the same issue and downgrading to 7.9.0 solved the issue.

**Comment by ImPDP** [comment]
@Edweis but downgrading to 7.9.0 doesnt solve my error. the warning is still showing in the console as atob is not found.

**Comment by rommelpe** [comment]
Hi y'all, thanks for the inputs. I've tried replicating this issue so I could escalate this to our Firestore folks, but can't seem to replicate it or I might be missing something here. Here's [my repro](https://github.com/rommelpe/react-native) [running in the browser](https://docs.expo.io/versions/v33.0.0/introduction/running-in-the-browser/) with the help of this [guide](https://docs.expo.io/versions/v36.0.0/guides/using-firebase/#using-expo-with-firestore).

If that's not the case, let me know so I can recreate a working repro and be able to reach out to right person.

**Comment by schmidt-sebastian** [comment]
Sorry to hear this. Since both 7.9.0 and 7.9.1 seem to be broken in different ways, for now I recommend using 7.8.2 with React Native.

**Comment by ImPDP** [comment]
I have tried this many lower versions still not working the same error is showing.

**Comment by thebrianchen** [comment]
I tried tinkering with @rommelpe's repro, and `atob()` seems to be defined, so I'm not getting the error. I'm currently investigating the issue, but in the meantime, does following the most-upvoted answer on this [stackoverflow post](https://stackoverflow.com/questions/42829838/react-native-atob-btoa-not-working-without-remote-js-debugging) (importing `base-64` and adding `atob` and `btoa` globals) work as a temporary fix? A recent change in 7.9 started relying on `atob`, which does not seem to be defined in the global scope in react-native.

Also, if someone could provide a minimal repro with the error, that would be greatly appreciated!

**Comment by ImPDP** [comment]
I have tried importing base64 globally. No change. This is the warning I got in the yellow box.
```

[Unhandled promise rejection: ReferenceError: Can't find variable: atob]

Stack trace:
  node_modules\@firebase\firestore\dist\index.cjs.js:23101:0 in <global>
  http://192.168.43.87:19001/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&minify=false&hot=false:129602:60 in fromBase64String
  http://192.168.43.87:19001/node_modules%5Cexpo%5CAppEntry.bundle?platform=android&dev=true&minify=false&hot=false:140371:55 in onMessage
  node_modules\@firebase\firestore\dist\index.cjs.js:1315:17 in BasePath.prototype.isImmediateParentOf
  node_modules\promise\setimmediate\core.js:37:14 in tryCallOne
  node_modules\promise\setimmediate\core.js:123:25 in setImmediate$argument_0
  node_modules\react-native\Libraries\Core\Timers\JSTimers.js:146:14 in _callTimer
  node_modules\react-native\Libraries\Core\Timers\JSTimers.js:194:17 in _callImmediatesPass
  node_modules\react-native\Libraries\Core\Timers\JSTimers.js:458:30 in callImmediates
  [native code]:null in callImmediates
  node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:407:6 in __callImmediates
  node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:143:6 in __guard$argument_0
  node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:384:10 in __guard
  node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:142:17 in __guard$argument_0
  [native code]:null in flushedQueue
  [native code]:null in callFunctionReturnFlushedQueue
```

**Comment by ImPDP** [comment]
Yes. Importing `base64` globally in the `app.js` fixed my problem. Thank you. But many of you suggested downgrading to lower version working. But I had downgraded to even version `6.5.0`. it showed the same results only. Anyway, global import seems to be the better solution as of now

**Comment by JoynerPerez1987** [comment]
Please do you have an example

**Comment by ImPDP** [comment]
Check out this [question](https://stackoverflow.com/questions/60361519/cant-find-a-variable-atob/60369040#60369040). I have posted my answer.

**Comment by NickFoden** [comment]
Using ` "firebase": "^7.9.0",` also worked for me. In meantime can we update the name of the issue, perhaps include bit about "unable to find variable atob"  to save future devs time locating this issue. Wasn't expecting firestore to be the culprit when hitting issues updating my RN app to latest deps today. 

Will this be resolved in the May release?

**Comment by carolinemcq** [comment]
This appears to be an issue again in 7.14.3

**Comment by johnorourke** [comment]
> This appears to still be an issue again in 7.14.3, only the workaround posted on StackOverflow no longer works. Is there another way to get around this?

I'm running 7.14.5 on React 16.13.1 and it's OK :)

**Comment by carolinemcq** [comment]
> > This appears to still be an issue again in 7.14.3, only the workaround posted on StackOverflow no longer works. Is there another way to get around this?
> 
> I'm running 7.14.5 on React 16.13.1 and it's OK :)

Ok - I'm a total noob at react native so I must be doing something wrong. Can you explain in a bit of detail how you solved this, please?

**Comment by johnorourke** [comment]
> > > This appears to still be an issue again in 7.14.3, only the workaround posted on StackOverflow no longer works. Is there another way to get around this?
> > 
> > 
> > I'm running 7.14.5 on React 16.13.1 and it's OK :)
> 
> Ok - I'm a total noob at react native so I must be doing something wrong. Can you explain in a bit of detail how you solved this, please?

Ignore me, sorry! I didn't read the issue title properly - I am using React web not native!

**Comment by schmidt-sebastian** [comment]
The workaround posted on StackOverflow worked for me, but it `atob` and `btoa` has to be registered before the Firebase SDK is loaded.

**Comment by NickFoden** [comment]
Does anyone know if this will be remedied soon or are react-native users planning to stay on version 7.9.0 ?

**Comment by schmidt-sebastian** [comment]
In review now: https://github.com/firebase/firebase-js-sdk/pull/3198

**Comment by banjianpan** [comment]
When i go the atob and btoa route, I get this error after about 1 min 7 secs: 
![103385295_552691125428358_2132763501409926224_n](https://user-images.githubusercontent.com/40034920/84849168-5a5e7700-b01a-11ea-9e6b-5963e9774de4.jpg)

does anyone know how to fix this? Also I saw that there was a pretty similar issue here: https://github.com/expo/expo/issues/8130 , so these issues might be related!

**Comment by luklew** [comment]
Upgraded to 7.15.2,

Getting error when attempting to talk to firestore:

firebase TypeError: util.base64.tI is not a function 

This only happens when running on iOS and Android, web works fine.

**Comment by schmidt-sebastian** [comment]
FYI: The current SDK release is mostly broken for RN. I suggest using 71,5.0 with the atob/bota Polyfill (7.15.2 **with** Polyfill should also work).

**Comment by mirfanahmed** [comment]
I upgraded to version 7.15.3 from 7.15.1 but still I cannot write data on cloud firestore in react native although it is written in the release notes that the problem is fixed. Have anyone tried the latest version?

**Comment by JoynerPerez1987** [comment]
yes, it is correct, it throws an error when you try to write

**Comment by mirfanahmed** [comment]
> yes, it is correct, it throws an error when you try to write

should I downgrade to other version which is working?

**Comment by JoynerPerez1987** [comment]
yes, it goes back to 7.9.0.

**Comment by mirfanahmed** [comment]
> yes, it goes back to 7.9.0.
> If you use Expo use this, in your package.json: "firebase": "7.9.0"

No I am using react native cli..

**Comment by JoynerPerez1987** [comment]
in that case I'm not sure which one to go back to

**Comment by luklew** [comment]
@JoynerPerez1987 
7.14.1 works ok for me

**Comment by JoynerPerez1987** [comment]
they just released a new version(7.15.4) and apparently there are still problems.[https://firebase.google.com/support/release-notes/js](url)

**Comment by schmidt-sebastian** [comment]
Re-opening until https://github.com/firebase/firebase-js-sdk/pull/3251 lands.
